akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -413,4 +414,226 @@ final class PrefixAndTail[T](n: Int) extends GraphStage[FlowShape[T, (immutable.
+      override def onPush(): Unit = {
+        val handler = new SubstreamHandler
+        val elem = grab(in)
+
+        decision match {
+          case SplitAfter if p(elem) ⇒
+            push(out, Source.single(elem))
+          // Next pull will come from the next substream that we will open
+          case _ ⇒
+            handler.firstElem = elem
+        }
+
+        handOver(handler)
+      }
+
+      override def onUpstreamFinish(): Unit = completeStage()
+
+    })
+
+    private def handOver(handler: SubstreamHandler): Unit = {
+      if (isClosed(out)) completeStage()
+      else {
+        val source = new SubSource[T](
+          timeout,
+          handler.onSubstreamRegister.invoke,
+          handler.onSubstreamPull.invoke,
+          handler.onSubstreamFinish.invoke)
+        substreamSource = source
+        substreamCancelled = false
+        setHandler(in, handler)
+
+        if (isAvailable(out)) {
+          push(out, Source.fromGraph(source))
+          substreamPushed = true
+        } else substreamPushed = false
+
+      }
+    }
+
+    override protected def onTimer(timerKey: Any): Unit =
+      if (substreamSource.materializationState.compareAndSet(NotMaterialized, TimedOut))
+        failStage(new SubscriptionTimeoutException(s"Substream Source has not been materialized in $timeout."))
+
+    private class SubstreamHandler extends InHandler {
+      var firstElem: T = null.asInstanceOf[T]
+      private var substream: SubSourceInterface[T] = null
+      private var pendingCompletion: MaterializationState = null
+      private var willCompleteAfterInitialElement = false
+      // The interpreter handles ignoring events for a port after it has been closed, but this is not a real
+      // port but an async side-channel. We need to ignore it ourselves.
+      private var ignoreAsync = false
+
+      private def waitingSubstreamRegistration: Boolean = substream eq null
+      private def hasInitialElement: Boolean = firstElem.asInstanceOf[AnyRef] ne null
+
+      // Substreams are always assumed to be pushable position when we enter this method
+      private def closeThis(handler: SubstreamHandler, currentElem: T): Unit = {
+        ignoreAsync = true
+        decision match {
+          case SplitAfter ⇒
+            if (!substreamCancelled) {
+              substream.pushSubstream(currentElem)
+              substream.completeSubstream()
+            }
+          case SplitBefore ⇒
+            handler.firstElem = currentElem
+            if (!substreamCancelled) substream.completeSubstream()
+        }
+      }
+
+      val onSubstreamPull = getAsyncCallback[Unit] { _ ⇒
+        if (!ignoreAsync) {
+          if (hasInitialElement) {
+            substream.pushSubstream(firstElem)
+            firstElem = null.asInstanceOf[T]
+            if (willCompleteAfterInitialElement) {
+              substream.completeSubstream()
+              ignoreAsync = true
+              completeStage()
```

Comment:
I’d prefer to keep it only where it has an effect, so as to not confuse the future reader.